File: Java/Leaderboard/src/main/java/tddmicroexercises/leaderboard/Leaderboard.java
package tddmicroexercises.leaderboard;

import java.util.*;
import java.util.stream.Collectors;

import static java.util.Map.Entry.comparingByValue;

public class Leaderboard {

  private final List<Race> races;

  public Leaderboard(Race... races) {
    this.races = Arrays.asList(races);
  }

  public Map<String, Integer> driverResults() {
    Map<String, Integer> results = new HashMap<>();
    for (Race race : this.races) {
      for (Driver driver : race.getResults()) {
        String driverName = race.getDriverName(driver);
        results.put(driverName, results.getOrDefault(driverName, 0) + race.getPoints(driver));
      }
    }
    return results;
  }

  public List<String> driverRankings() {
    ArrayList<Map.Entry<String, Integer>> entries = new ArrayList<>(driverResults().entrySet());
    entries.sort(comparingByValue(Comparator.reverseOrder()));
    return entries.stream().map(Map.Entry::getKey).collect(Collectors.toList());
  }
}

File: kotlin/Leaderboard/src/main/kotlin/tddmicroexercises/leaderboard/Leaderboard.kt
package tddmicroexercises.leaderboard

import java.util.*

class Leaderboard(vararg races: Race) {

    private val races: List<Race>

    init {
        this.races = arrayListOf(*races)
    }

    fun driverResults(): Map<String, Int> {
        val results = HashMap<String, Int>()
        for (race in this.races) {
            for (driver in race.results) {
                val driverName = race.getDriverName(driver)
                val points = race.getPoints(driver)
                if (results.containsKey(driverName)) {
                    results[driverName!!] = results[driverName]!!.plus(points)
                } else {
                    results[driverName!!] = 0 + points
                }
            }
        }
        return results
    }

    fun driverRankings(): List<String> {
        val results = driverResults()
        val resultsList = ArrayList(results.keys)
        Collections.sort(resultsList, DriverByPointsDescendingComparator(results))
        return resultsList
    }

    private class DriverByPointsDescendingComparator constructor(private val results: Map<String, Int>) :
        Comparator<String> {

        override fun compare(driverName1: String, driverName2: String): Int {
            return -results[driverName2]?.let { results[driverName1]!!.compareTo(it) }!!
        }
    }

}
